sympy/polys/polyfuncs.py

```diff
@@ -232,31 +233,39 @@ def interpolate(data, x):
     >>> interpolate({-1: 2, 1: 2, 2: 5}, x)
     x**2 + 1
 
+    If the interpolation is going to be used only once then the
+    value of interest can be passed instead of passing a symbol:
+
+    >>> interpolate([1, 4, 9], 5)
+    25
+
+    Symbolic coordinates are also supported:
+
+    >>> [(i,interpolate((a, b), i)) for i in range(1, 4)]
+    [(1, a), (2, b), (3, -a + 2*b)]
     """
     n = len(data)
-    poly = None
 
     if isinstance(data, dict):
+        if x in data:
+            return S(data[x])
         X, Y = list(zip(*data.items()))
-        poly = interpolating_poly(n, x, X, Y)
     else:
         if isinstance(data[0], tuple):
             X, Y = list(zip(*data))
-            poly = interpolating_poly(n, x, X, Y)
+            if x in X:
+                return S(Y[X.index(x)])
         else:
+            if x in range(1, n + 1):
+                return S(data[x - 1])
             Y = list(data)
+            X = list(range(1, n + 1))
 
-            numert = Mul(*[(x - i) for i in range(1, n + 1)])
-            denom = -factorial(n - 1) if n%2 == 0 else factorial(n - 1)
-            coeffs = []
-            for i in range(1, n + 1):
-                coeffs.append(numert/(x - i)/denom)
-                denom = denom/(i - n)*i
-
-            poly = Add(*[coeff*y for coeff, y in zip(coeffs, Y)])
-
-    return poly.expand()
-
+    try:
+        return interpolating_poly(n, x, X, Y).expand()
+    except ValueError:
+        d = Dummy()
+        return interpolate(data, d).subs(d, x)
```

Comment:
Perhaps this could also call `interpolating_poly` to avoid having to repeat the construction of `X` and `Y`.